pmd-scala/src/main/java/net/sourceforge/pmd/lang/scala/ast/ScalaWrapperNode.java

```diff
@@ -0,0 +1,155 @@
+/**
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.scala.ast;
+
+import java.util.Objects;
+
+import net.sourceforge.pmd.lang.ast.AbstractNode;
+import net.sourceforge.pmd.lang.ast.Node;
+import net.sourceforge.pmd.lang.scala.ScalaParser;
+
+import scala.meta.Tree;
+import scala.meta.inputs.Position;
+
+/**
+ * The Java-wrapper for Scala Nodes to allow interoperability with PMD.
+ */
+public class ScalaWrapperNode extends AbstractNode implements ScalaNode {
+    private Tree node;
+    private ScalaParser parser;
+
+    /**
+     * Create a new Java Wrapper around a Scala node.
+     * 
+     * @param scalaParser
+     *            the ScalaParser used to generate the node
+     * @param scalaNode
+     *            the scalaNode node to wrap
+     */
+    public ScalaWrapperNode(ScalaParser scalaParser, Tree scalaNode) {
+        super(0);
+        this.parser = scalaParser;
+        this.node = scalaNode;
+        Position pos = node.pos();
+        beginLine = pos.startLine() + 1;
+        endLine = pos.endLine() + 1;
+        beginColumn = pos.startColumn() + 1;
+        endColumn = pos.endColumn() + 1;
+    }
+
+    @Override
+    public Tree getNode() {
+        return node;
+    }
+
+    @Override
+    public Object accept(ScalaParserVisitor visitor, Object data) {
+        return visitor.visit(this, data);
+    }
+
+    @Override
+    public Object childrenAccept(ScalaParserVisitor visitor, Object data) {
+        int numChildren = jjtGetNumChildren();
+        for (int i = 0; i < numChildren; ++i) {
+            jjtGetChild(i).accept(visitor, data);
+        }
+        return data;
+    }
+
+    @Override
+    public String getXPathNodeName() {
+        return "AST" + node.productPrefix().replace(".", "") + "Node";
+    }
+
+    @Override
+    public void jjtClose() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void jjtSetParent(Node parent) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public ScalaWrapperNode jjtGetParent() {
+        if (node.parent().isEmpty()) {
+            return null;
+        }
+        return parser.wrapNode(node.parent().get());
+    }
+
+    @Override
+    public void jjtAddChild(Node child, int index) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void jjtSetChildIndex(int index) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public int jjtGetChildIndex() {
+        int idx = node.parent().get().children().indexOf(node);
+        if (idx == -1) {
+            throw new IllegalStateException("This node is not a child of its parent: " + node);
+        }
+        return idx;
+    }
+
+    @Override
+    public ScalaWrapperNode jjtGetChild(int index) {
+        return parser.wrapNode(node.children().apply(index));
```

Comment:
This wrapping on the fly is an interesting technique. In other languages (Apex, JavaScript) we wrap the nodes eagerly and we have a PMD specific node class for each original AST node type. That way, the visitors are more expressive and the different node types can have more easily different behaviour (e.g. `getImage()`) and a more expressive API (`getCondition()`). This helps later on in writing Java-based rules and maybe XPath-based rules as well.
I don't know yet, whether we should do it like this or not...

Reply:
This is a good point. I tend to use XPath more than the Java API and this was a simple way to get there quickly.
I'm working on converting the Scala Tree Nodes to Java Nodes, but there are ~250, so it's taking a bit.